### ISSUE TITLE
App crash when implementing Cloud Saves

### ISSUE THREAD

**Issue by EkoEkey** [playgameservices/play-games-plugin-for-unity#1905]
I have an unity app in beta version published on Google Play Console.
Currently I have published it to beta testers and I think I have successfully configurated Google Play Games Plugin in Unity and Setting in Google Play Development Console (I'm sure that i have activated "save games" in Development Console) and I think I've set everything up correctly (SHA-1 numbers are already all the same).

**With this code everything works perfectly:**
`using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using GooglePlayGames;
using GooglePlayGames.BasicApi;
using UnityEngine.SocialPlatforms;

public class PlayManager : MonoBehaviour {
    void Start () {
       PlayGamesClientConfiguration config = new PlayGamesClientConfiguration.Builder()
            .Build();

       PlayGamesPlatform.InitializeInstance(config);

       PlayGamesPlatform.DebugLogEnabled = true;

       PlayGamesPlatform.Activate();

       if (!Social.localUser.authenticated)
          Social.localUser.Authenticate(success => { });
    }`

**But when i implemented "EnabledSaveGames" in PlayGamesClientConfiguration with this code app crash after Play Games login.**

`using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using GooglePlayGames;
using GooglePlayGames.BasicApi;
using UnityEngine.SocialPlatforms;

public class PlayManager : MonoBehaviour {
    void Start () {
       PlayGamesClientConfiguration config = new PlayGamesClientConfiguration.Builder()
           .EnableSavedGames()
           .Build();

       PlayGamesPlatform.InitializeInstance(config);

       PlayGamesPlatform.DebugLogEnabled = true;

       PlayGamesPlatform.Activate();

       if (!Social.localUser.authenticated)
          Social.localUser.Authenticate(success => { });
    }`

[Adb Logcat Log](https://pastebin.com/RksQPQ6C)
[Console Logcat Log](https://pastebin.com/652v8Dkf)

I really don't know what are all that "art" errors, but I don't think that are related to Google Play.
I don't understand this warning:  "Bad call: specified package com.google.android.play.games under uid 10219 but it is really 10118."

**Unity Version: 5.5.1f1
Google Play Games Plugin version: 0.9.40
SDK Manager Google Play Games Service version 43**

Hope that somebody could help me because I'm in this situation since two days.

**[SOLVED]**
Possible solutions:
-My device was not in the supported one in the Google Play Development Console.
-I have published App Services.

**Comment by jsaraiva** [comment]
Hi!

There is a line in your console logcat:
09-05 15:53:47.564: E/GamesNativeSDK(27343): Exception in com/google/android/gms/games/snapshot/Snapshots.load: java.lang.IllegalStateException: Cannot use snapshots without enabling the 'Saved Game' feature in the Play console.

You said that you activated Saved Games in the Play Console. However, does the project have the Drive API activated?
To check, go to Game Services -> <Your game> -> Game details . This is the page where you (de)activate Saved Games; scroll to the bottom of the page, to the "API Console project" section, and check if the "APIs required for Saved Games to work" entry has a green checkmark next to it. If it doesn't, you have to go to Google's Development Console and assign that API to your project (just follow the project link in the API Console project session).

Regarding the other lines, they seem to be permissions-related warnings (as per the W letter at the start).

Hope this helps!

Cheers!

**Comment by EkoEkey** [comment]
Hi @jsaraiva , thank you for your comment.
I have checked APIs required for "Saved Games to work" in game details and it have the green checkmark.
I don't think the problem is related to API, that have either the green checkmark.

**Comment by eddiewillman** [comment]
I have the same issue, but **only for certain Google Play accounts**. Our app is live on Google Play. The Save Games feature has been working fine while we were using the Prime31 GPG plugin. When I switched to this plugin it seems to work for most of the existing users/accounts. But for a few users the app crashes a second or so after the authentication process starts. The following is written to the log as it crashes (see attached files for complete logs):

```
Unity: [Play Games Plugin DLL] 09/06/17 15:20:02 +09:00 DEBUG: Starting Auth Transition. Op: SIGN_IN status: VALID
GamesNativeSDK: Exception in com/google/android/gms/games/snapshot/Snapshots.load: java.lang.IllegalStateException: Cannot use snapshots without enabling the 'Saved Game' feature in the Play console.
GamesNativeSDK: Exception in com/google/android/gms/games/achievement/Achievements.load: java.lang.SecurityException: Not signed in when calling API.
GamesNativeSDK: Could not automatically seed snapshot cache.
art: art/runtime/java_vm_ext.cc:410] JNI DETECTED ERROR IN APPLICATION: can't call void com.google.android.gms.common.api.PendingResult.setResultCallback(com.google.android.gms.common.api.ResultCallback) on null object
```

Please not that for users that doesn't experience the crash the "Cannot use snapshots ..." message is not written to the log.

The crash happens both for a completely new Google account and for a real user account which was able to authenticate with GPG previously when we were using the Prime31 plugin. Adding the newly created Google account as a test user for the app and game services had no effect.

To test this I created a new Unity project, imported GPG, added a button to the scene, and added a script containing:

```
void Start () {
    PlayGamesPlatform.InitializeInstance(new PlayGamesClientConfiguration.Builder().EnableSavedGames().Build());
    PlayGamesPlatform.DebugLogEnabled = true;
    PlayGamesPlatform.Activate();
}

public void OnClick() {
    Social.localUser.Authenticate(success => {});
}
```

The apk was built with the same **bundle identifier**, **keystore**, and **version code** as the build currently live on Google Play. The live version still uses the Prime31 plugin and has functioning save games. Possibly unrelated to this the Prime31 plugin triggers a modal for new users asking them to create a Google+ profile during the authentication process.

**"Save Games" is enabled** under game details on the game services section on Google Play Console.

Under "API Console project" **the game links to an API console project**. Green check marks are displayed for "APIs required for basic Games Services to work" and "APIs required for Saved Games to work".

I get the same crash with both version **0.9.28** and **0.9.40**. As a side note, I wished to use version 0.9.28 to not have the app ask users with older versions of Google Play Services to update.

The Android manifest packaged in the APK seems fine with the correct GMS app ID etc. My test project doesn't provide a custom manifest.

- Unity 5.5.4f1
- Google Play Games Plugin 0.9.40
- play-services aar files 11.0.4
- support aar files 25.2.0
- Android SDK platform 7.1.1 (Nougat)
- Android SDK Build-Tools 25.0.3
- Android SDK Platform-Tools 25.0.5
- Android SDK Tools 25.2.5
- Google Play services 43

[logcat-failure.txt](https://github.com/playgameservices/play-games-plugin-for-unity/files/1280535/logcat-failure.txt)
[logcat-success.txt](https://github.com/playgameservices/play-games-plugin-for-unity/files/1280536/logcat-success.txt)

**Comment by EkoEkey** [comment]
It's been 48 hours since i have activated "Save Games" in GP Development Console I still have app crashes.
I have re-done a logcat debug and the most significative errors i got are these (Same as before):

```
09-06 10:29:07.022: E/GamesNativeSDK(992): Exception in com/google/android/gms/games/snapshot/Snapshots.load: java.lang.IllegalStateException: Cannot use snapshots without enabling the 'Saved Game' feature in the Play console.

09-06 10:29:07.188: E/GamesNativeSDK(992): Exception in com/google/android/gms/games/achievement/Achievements.load: java.lang.SecurityException: Not signed in when calling API.

09-06 10:29:07.235: A/art(992): art/runtime/check_jni.cc:65] JNI DETECTED ERROR IN APPLICATION: obj == null
```

I don't understand these error, i have activated "Save games" more than 2 days ago.
If I toggle .EnableSavedGames() in PlayGamesClientConfiguration everything works fine.

Someone have some suggest?

**Comment by eddiewillman** [comment]
> @EkoEkey If I toggle .EnableSavedGames() in PlayGamesClientConfiguration everything works fine.

You must do this to initialize the GPG plugin with the save games feature enabled on the client. Otherwise the save game feature will not work.

**Comment by EkoEkey** [comment]
Hi @eddiewillman thank you for your comments.
I know that i have to inizitalize  .EnableSavedGames() and i do it but when i do that I have App crashes.

Personally i dind't have any experience with Prime31 GPG plugin.
I can say that after 2 days since I have activated "Save Games" featured in GP Development Console i still have app crashes that look the same as yours when .EnableSavedGames() is implemented in the client.

These are the most signicative:
```
09-06 11:56:27.031: E/GamesNativeSDK(9221): Exception in com/google/android/gms/games/snapshot/Snapshots.load: java.lang.IllegalStateException: Cannot use snapshots without enabling the 'Saved Game' feature in the Play console.
09-06 11:56:27.158: E/GamesNativeSDK(9221): Exception in com/google/android/gms/games/achievement/Achievements.load: java.lang.SecurityException: Not signed in when calling API.
09-06 11:56:27.215: A/art(9221): art/runtime/check_jni.cc:65] JNI DETECTED ERROR IN APPLICATION: obj == null
```
I hope that somebody can help.

**Comment by EkoEkey** [comment]
Today I have tried to publish the Game services with "Save games" setted as active. 
After some hours I still have the same error.
`09-06 11:56:27.031: E/GamesNativeSDK(9221): Exception in com/google/android/gms/games/snapshot/Snapshots.load: java.lang.IllegalStateException: Cannot use snapshots without enabling the 'Saved Game' feature in the Play console.`

I'll update this comment tomorrow to see if something is changed.

**Comment by EkoEkey** [comment]
**[SOLVED]** Today i have take a look to the supported devices in the Google Play Console and i noticed that my phone isn't in there. So i think that the main problem of all was that.
Thank you all for your help.

**Comment by eddiewillman** [comment]
Update on the similar issue I had that only affected certain Google Play accounts. It was caused by us having unpublished changes to the app in the Google Play game services dashboard. Once we published those all users could authenticate without issue within a few hours.
Most if not all the accounts we use internally when testing are registered as testers. Since testers use any saved but unpublished changes these accounts did not experience the issue.
